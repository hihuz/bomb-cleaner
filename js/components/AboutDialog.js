import React from 'react';

const AboutDialog = ({ closeAboutDialog, opened }) => (
  <div className={`dialog${opened?' opened':''}`}>
    <button className='dialog-close' onClick={closeAboutDialog}>
      <i className='icon-close'></i>
    </button>
    <p>
      This is a MineSweeper game implemented with React and Redux.
      <br />
      <br />
      You can play it in the same way as the original MineSweeper.
    </p>
    <ul>
      <li>Left Click to open a tile</li>
      <li>Right Click to "flag" a tile</li>
      <li>You'll lose if you open a mine</li>
      <li>You'll win if you open all mine-free tiles</li>
    </ul>
    <p className='centered-text'>
      Good luck !
      <br />
      <br />
      <a target="_blank" href="https://twitter.com/hihuz">@hihuz</a>
    </p>
  </div>
);
export default AboutDialog;
