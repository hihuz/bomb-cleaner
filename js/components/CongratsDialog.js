import React from 'react';
import AddHSForm from './AddHSForm';

export function isHighScore(mode, time, highScores) {
  const worseTimes = highScores[mode].filter(hs => time <= hs.time).length;
  return highScores[mode].length < 5 || worseTimes > 0;
}

const CongratsDialog = ({ closeDialog, time, mode, highScores }) => (
  <div>
    <button className="dialog-close" onClick={closeDialog}>
      <i className="icon-close" />
    </button>
    <p className="congrats-para">
      Congratulations !
    </p>
    {mode && time && highScores && isHighScore(mode, time, highScores) ? <AddHSForm
      mode={mode}
      time={time}
      highScores={highScores}
      closeDialog={closeDialog}
    /> : ''}
  </div>
);

export default CongratsDialog;
