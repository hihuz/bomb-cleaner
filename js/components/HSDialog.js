import React from 'react';

const HSDialog = ({ closeHSDialog, opened }) => (
  <div className={`dialog${opened?' opened':''}`}>
    <button className="dialog-close" onClick={closeHSDialog}>x</button>
    lu
  </div>
);

export default HSDialog;
