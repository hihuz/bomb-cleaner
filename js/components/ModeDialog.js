import React from 'react';
import { connect } from 'react-redux';
import { setMode } from '../actions/actionCreators';

class ModeDialog extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mode: this.props.mode,
      width: this.props.width,
      height: this.props.height,
      bombs: this.props.bombs
    };
    this.handleCloseClick = this.handleCloseClick.bind(this);
    this.setLocalEasyMode = this.setLocalEasyMode.bind(this);
    this.setLocalMediumMode = this.setLocalMediumMode.bind(this);
    this.setLocalHardMode = this.setLocalHardMode.bind(this);
    this.setLocalCustomMode = this.setLocalCustomMode.bind(this);
    this.verifyWidthValue = this.verifyWidthValue.bind(this);
    this.setLocalWidth = this.setLocalWidth.bind(this);
    this.verifyHeightValue = this.verifyHeightValue.bind(this);
    this.setLocalHeight = this.setLocalHeight.bind(this);
    this.verifyBombsValue = this.verifyBombsValue.bind(this);
    this.setLocalBombs = this.setLocalBombs.bind(this);
    this.pushNewMode = this.pushNewMode.bind(this);
  }
  setLocalEasyMode() {
    this.setState({
      mode: 'easy',
      width: 9,
      height: 9,
      bombs: 10
    });
  }
  setLocalMediumMode() {
    this.setState({
      mode: 'medium',
      width: 16,
      height: 16,
      bombs: 40
    });
  }
  setLocalHardMode() {
    this.setState({
      mode: 'hard',
      width: 30,
      height: 16,
      bombs: 99
    });
  }
  setLocalCustomMode() {
    this.setState({ mode: 'custom' });
  }
  setLocalWidth(e) {
    this.setState({ width: e.target.value });
  }
  setLocalHeight(e) {
    this.setState({ height: e.target.value });
  }
  setLocalBombs(e) {
    this.setState({ bombs: e.target.value });
  }
  verifyWidthValue(e) {
    const value = Number(e.target.value);
    if (isNaN(value) || value < 9) {
      this.setState({ width: 9 });
    } else if (value > 24) {
      this.setState({ width: 30 });
    }
  }
  verifyHeightValue(e) {
    const value = Number(e.target.value);
    if (isNaN(value) || value < 9) {
      this.setState({ height: 9 });
    } else if (value > 30) {
      this.setState({ height: 30 });
    }
  }
  verifyBombsValue(e) {
    const value = Number(e.target.value);
    const size = this.state.width * this.state.height;
    if (isNaN(value) || value < 10) {
      this.setState({ bombs: 10 });
    } else if (value > 668 && size > 668) {
      this.setState({ bombs: 668 });
    } else if (value >= size) {
      this.setState({ bombs: size - 1 });
    }
  }
  handleCloseClick() {
    this.setState({ mode: this.props.mode });
    this.props.closeDialog();
  }
  pushNewMode() {
    this.props.dispatch(setMode(this.state));
    this.props.closeDialog();
  }
  render() {
    return (
      <div>
        <button className="dialog-close" onClick={this.handleCloseClick}>
          <i className="icon-close" />
        </button>
        <button
          className="mode-dialog-button"
          disabled={this.state.mode === 'easy'}
          onClick={this.setLocalEasyMode}
        >Easy</button>
        <button
          className="mode-dialog-button"
          disabled={this.state.mode === 'medium'}
          onClick={this.setLocalMediumMode}
        >Medium</button>
        <button
          className="mode-dialog-button"
          disabled={this.state.mode === 'hard'}
          onClick={this.setLocalHardMode}
        >Hard</button>
        <button
          className="mode-dialog-button"
          disabled={this.state.mode === 'custom'}
          onClick={this.setLocalCustomMode}
        >Custom:</button>
        <div>
          <label htmlFor="width-input">Width (9-30): </label>
          <input
            className="mode-input"
            maxLength="2"
            disabled={this.state.mode !== 'custom'}
            value={this.state.width}
            type="text"
            id="width-input"
            onBlur={this.verifyWidthValue}
            onChange={this.setLocalWidth}
          />
        </div>
        <div>
          <label htmlFor="height-input">Height (9-30): </label>
          <input
            className="mode-input"
            maxLength="2"
            disabled={this.state.mode !== 'custom'}
            value={this.state.height}
            type="text"
            id="height-input"
            onBlur={this.verifyHeightValue}
            onChange={this.setLocalHeight}
          />
        </div>
        <div>
          <label htmlFor="bombs-input">Mines (10-668): </label>
          <input
            className="mode-input"
            maxLength="3"
            disabled={this.state.mode !== 'custom'}
            value={this.state.bombs}
            type="text"
            id="bombs-input"
            onBlur={this.verifyBombsValue}
            onChange={this.setLocalBombs}
          />
        </div>
        <button className="mode-validate" onClick={this.pushNewMode}>OK</button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  mode: state.mode,
  width: state.grid.width,
  height: state.grid.height,
  bombs: state.grid.bombs
});

export const Plain = ModeDialog;
export default connect(mapStateToProps)(ModeDialog);
