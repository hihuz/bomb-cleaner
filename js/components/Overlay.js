import React from 'react';
import ModeDialog from './ModeDialog';
import HSDialog from './HSDialog';
import AboutDialog from './AboutDialog';
import CongratsDialog from './CongratsDialog';

const Overlay = ({
  modeDialogOpened,
  closeModeDialog,
  hsDialogOpened,
  highScores,
  closeHSDialog,
  aboutDialogOpened,
  closeAboutDialog,
  congratsDialogOpened,
  closeCongratsDialog,
  time,
  mode
}) => (
  <div className="overlay">
    <ModeDialog
      opened={modeDialogOpened}
      closeModeDialog={closeModeDialog}
    />
    <HSDialog
      opened={hsDialogOpened}
      highScores={highScores}
      closeHSDialog={closeHSDialog}
    />
    <AboutDialog
      opened={aboutDialogOpened}
      closeAboutDialog={closeAboutDialog}
    />
    <CongratsDialog
      time={time}
      mode={mode}
      highScores={highScores}
      opened={congratsDialogOpened}
      closeCongratsDialog={closeCongratsDialog}
    />
  </div>
);

export default Overlay;
