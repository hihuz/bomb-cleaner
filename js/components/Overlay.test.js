import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Overlay from './Overlay';
import ModeDialog from './ModeDialog';
import HSDialog from './HSDialog';
import AboutDialog from './AboutDialog';
import CongratsDialog from './CongratsDialog';

test('Snapshot', () => {
  const component = shallow(<Overlay />);
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});

test('Should call the closeDialog prop when the overlay is clicked', () => {
  const clickFn = jest.fn();
  const component = shallow(<Overlay closeDialog={clickFn} />);
  component.simulate('click', { target: '', currentTarget: '' });
  expect(clickFn).toHaveBeenCalledTimes(1);
});

test('Should not render any dialog component if dialogOpened props is false', () => {
  const component = shallow(<Overlay dialogOpened={false} />);
  expect(component.find('.dialog').children('div').length).toEqual(0);
  expect(component.find(ModeDialog).length).toEqual(0);
  expect(component.find(HSDialog).length).toEqual(0);
  expect(component.find(AboutDialog).length).toEqual(0);
  expect(component.find(CongratsDialog).length).toEqual(0);
});

test('Should render a ModeDialog component if dialogOpened props is "mode"', () => {
  const clickFn = jest.fn();
  const component = shallow(<Overlay closeDialog={clickFn} dialogOpened={'mode'} />);
  const mode = component.find(ModeDialog);
  expect(mode.length).toEqual(1);
  expect(mode.props().closeDialog).toEqual(clickFn);

  expect(component.find(HSDialog).length).toEqual(0);
  expect(component.find(AboutDialog).length).toEqual(0);
  expect(component.find(CongratsDialog).length).toEqual(0);
});

test('Should render a HSDialog component if dialogOpened props is "hs"', () => {
  const clickFn = jest.fn();
  const component = shallow(<Overlay
    closeDialog={clickFn}
    dialogOpened={'hs'}
    highScores={'foo'}
  />);
  const hs = component.find(HSDialog);
  expect(hs.length).toEqual(1);
  expect(hs.props().closeDialog).toEqual(clickFn);
  expect(hs.props().highScores).toEqual('foo');

  expect(component.find(ModeDialog).length).toEqual(0);
  expect(component.find(AboutDialog).length).toEqual(0);
  expect(component.find(CongratsDialog).length).toEqual(0);
});

test('Should render a AboutDialog component if dialogOpened props is "about"', () => {
  const clickFn = jest.fn();
  const component = shallow(<Overlay closeDialog={clickFn} dialogOpened={'about'} />);
  const about = component.find(AboutDialog);
  expect(about.length).toEqual(1);
  expect(about.props().closeDialog).toEqual(clickFn);

  expect(component.find(ModeDialog).length).toEqual(0);
  expect(component.find(HSDialog).length).toEqual(0);
  expect(component.find(CongratsDialog).length).toEqual(0);
});

test('Should render a CongratsDialog component if dialogOpened props is "congrats"', () => {
  const clickFn = jest.fn();
  const component = shallow(<Overlay
    closeDialog={clickFn}
    dialogOpened={'congrats'}
    highScores={'foo'}
    time={90}
    mode={'hard'}
  />);
  const congrats = component.find(CongratsDialog);
  expect(congrats.length).toEqual(1);
  expect(congrats.props().highScores).toEqual('foo');
  expect(congrats.props().time).toEqual(90);
  expect(congrats.props().mode).toEqual('hard');

  expect(component.find(ModeDialog).length).toEqual(0);
  expect(component.find(HSDialog).length).toEqual(0);
  expect(component.find(AboutDialog).length).toEqual(0);
});
