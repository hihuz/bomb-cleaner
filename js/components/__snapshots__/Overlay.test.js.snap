exports[`test Snapshot 1`] = `
<div
  className="overlay"
  onClick={[Function]}
  onKeyDown={[Function]}>
  <div
    className="dialog"
    tabIndex={-1} />
</div>
`;
